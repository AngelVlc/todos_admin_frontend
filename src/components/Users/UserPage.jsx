import React, { useEffect, useState } from 'react';
import { Link, useParams, useHistory } from 'react-router-dom';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import axios from 'axios';
import * as Yup from 'yup';

export const UserPage = (props) => {
    let { userId } = useParams();
    let history = useHistory();

    const initialState = {
        title: '',
        submintBtnText: '',
        submitUrl: 'users',
        isNew: true,
        name: '',
        isAdmin: ''
    }

    const [pageState, setPageState] = useState(initialState);

    useEffect(() => {
        const getExistingUser = async () => {
            const res = await axios.get(`users/${userId}`)
            const user = res.data

            let title = `Edit user '${user.name}'`
            if (user.isAdmin) {
                title = `Edit admin user '${user.name}'`
            }

            setPageState({
                title: title,
                submintBtnText: 'SAVE',
                submitUrl: `users/${userId}`,
                isNew: false,
                name: user.name,
                isAdmin: user.isAdmin ? 'yes' : 'no'
            });
        }
        if (userId) {
            getExistingUser();
        } else {
            setPageState({
                title: 'New user',
                submintBtnText: 'CREATE',
                submitUrl: 'users',
                isNew: true,
                name: '',
                isAdmin: 'no'
            });
        }
    }, [userId]);

    const onSubmit = async ({ name, isAdmin, password, confirmPassword }) => {
        const body = {
            name,
            password,
            confirmPassword,
            isAdmin: isAdmin === 'yes' ? true : false
        };
        let res;
        if (pageState.isNew) {
            res = await axios.post(pageState.submitUrl, body);
        } else {
            res = await axios.put(pageState.submitUrl, body);
        }
        history.push(`/users/${res.data.id}/edit`);
    }

    return (
        <div className="container">
            <h3 className="title">{pageState.title}</h3>
            <nav className="breadcrumb" aria-label="breadcrumbs">
                <ul>
                    <li><Link to={`/`}>Home</Link></li>
                    <li><Link to={`/users`}>Users</Link></li>
                    <li className="is-active"><Link aria-current="page" to={`/users/${userId}`}>{pageState.isNew ? 'new' : pageState.name}</Link></li>
                </ul>
            </nav>
            <Formik
                enableReinitialize={true}
                initialValues={{
                    name: pageState.name,
                    isAdmin: pageState.isAdmin,
                    password: '',
                    confirmPassword: ''
                }}
                validationSchema={
                    Yup.object({
                        name: Yup.string()
                            .required('Required')
                    })
                }
                onSubmit={onSubmit}>
                <Form>
                    <div className="field">
                        <label className="label" htmlFor="name">Name</label>
                        <div className="control">
                            <Field name="name" type="text" data-testid="name" />
                        </div>
                        <p className="help is-danger" data-testid="userNameErrors">
                            <ErrorMessage name="name" />
                        </p>
                    </div>

                    <div className="field">
                        <label className="label" htmlFor="password">Password</label>
                        <div className="control">
                            <Field name="password" type="password" data-testid="password" />
                        </div>
                    </div>

                    <div className="field">
                        <label className="label" htmlFor="confirmPassword">Confirm Password</label>
                        <div className="control">
                            <Field name="confirmPassword" type="password" data-testid="confirmPassword" />
                        </div>
                    </div>

                    <div className="field">
                        <label className="label" htmlFor="isAdmin">Is Admin</label>
                        <div className="select">
                            <Field name="isAdmin" as="select" data-testid="isAdmin">
                                <option value="yes">Yes</option>
                                <option value="no">No</option>
                            </Field>
                        </div>
                    </div>

                    <div className="field is-grouped">
                        <p className="control">
                            <button className="button" data-testid="submit" type="submit">{pageState.submintBtnText}</button>
                        </p>
                        <p className="control">
                            <button className="button" data-testid="cancel" type="button" onClick={() => history.push('/users')}>CANCEL</button>
                        </p>
                        {!pageState.isNew &&
                            <p className="control">
                                <button className="button is-danger" data-testid="delete" type="button" onClick={() => history.push(`/users/${userId}/delete`)}>DELETE</button>
                            </p>
                        }
                    </div>
                </Form>
            </Formik>
        </div>
    )
}