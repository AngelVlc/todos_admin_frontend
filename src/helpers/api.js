import { requestPending, requestDone, requestFailed } from '../actions'

const backendUrl = process.env.REACT_APP_BACKEND_URL || 'http://localhost:5001'

export const doGetToken = (loginDto, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(loginDto)
        };
        fetch(`${backendUrl}/auth/token`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (!res.ok) {
                    res.text().then(txt => reject(txt));
                    return
                }
                res.json().then(obj => {
                    resolve(obj);
                })
            })
            .catch(error => {
                requestsDispatch(requestDone());
                reject(error.message);
            })
    });
}

const getHeaders = (method, token) => {
    return {
        method: method,
        headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` },
    };
}

export const doGet = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('GET', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    res.json().then(obj => {
                        resolve(obj);
                    })
                } else {
                    res.text().then(txt => {
                        requestsDispatch(requestFailed(txt));
                        reject(txt)
                    });
                }
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doDelete = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('DELETE', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    resolve();
                    return
                }
                res.text().then(txt => {
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doPost = (endpoint, body, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('POST', token);
        requestOptions.body = JSON.stringify(body);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                res.text().then(txt => {
                    if (res.ok) {
                        resolve(txt);
                        return
                    }
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}
