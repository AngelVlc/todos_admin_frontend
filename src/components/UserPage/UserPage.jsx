import React, { useContext } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { useFormik } from 'formik';
import { AppContext } from '../../contexts/AppContext';
import { doPost, doPut } from '../../helpers/api';
import * as Yup from 'yup';

export const UserPage = (props) => {
    const { auth, requestsDispatch } = useContext(AppContext)
    let { userId } = useParams();
    let history = useHistory();

    const state = props.location.state;
    var returnUrl = '/'
    var title = 'New user'
    var submintBtnText = 'CREATE'
    var userName = ''
    var isAdmin = false
    var postUrl = 'users'
    if (state) {
        returnUrl = state.returnUrl;
        title = `Edit user '${state.userName}' (is admin: ${state.isAdmin})`
        submintBtnText = 'SAVE';
        ({ userName, isAdmin } = state);
        postUrl = `users/${userId}`;
    }

    const formik = useFormik({
        initialValues: {
            name: userName,
            isAdmin: isAdmin,
            newPassword: '',
            confirmNewPassword: ''
        },
        validationSchema: Yup.object({
            name: Yup.string()
                .required('Required')
        }),
        onSubmit: values => {
            if (!state) {
                doPost(postUrl, values, auth.info.token, requestsDispatch)
                    .then(() => {
                        history.push(returnUrl);
                    })
            } else {
                doPut(postUrl, values, auth.info.token, requestsDispatch)
                    .then(() => {
                        history.push(returnUrl);
                    })
            }
        }
    });

    return (
        <>
            <h4>{title}</h4>
            <form onSubmit={formik.handleSubmit}>
                <label htmlFor="name">Name</label>
                <input
                    id="name"
                    name="name"
                    onChange={formik.handleChange}
                    onBlur={formik.handleBlur}
                    value={formik.values.name}
                />
                {formik.touched.name && formik.errors.name ? (
                    <div>{formik.errors.name}</div>
                ) : null}
                <label htmlFor="newPassword">Password</label>
                <input
                    type="password"
                    id="newPassword"
                    name="newPassword"
                    onChange={formik.handleChange}
                    onBlur={formik.handleBlur}
                    value={formik.values.newPassword}
                />
                {formik.touched.newPassword && formik.errors.newPassword ? (
                    <div>{formik.errors.newPassword}</div>
                ) : null}
                <label htmlFor="confirmNewPassword">Confirm Password</label>
                <input
                    type="password"
                    id="confirmNewPassword"
                    name="confirmNewPassword"
                    onChange={formik.handleChange}
                    onBlur={formik.handleBlur}
                    value={formik.values.confirmNewPassword}
                />
                <label htmlFor="isAdmin">Is Admin</label>
                <input
                    type="checkbox"
                    id="isAdmin"
                    name="isAdmin"
                    onChange={formik.handleChange}
                    onBlur={formik.handleBlur}
                    checked={formik.values.isAdmin}
                />
                <button type="submit">{submintBtnText}</button>
                <button onClick={() => history.push(returnUrl)}>CANCEL</button>
            </form>
        </>
    )
}