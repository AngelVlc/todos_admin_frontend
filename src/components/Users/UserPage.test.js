import React from 'react'
import { render, cleanup, fireEvent, wait } from '@testing-library/react'
import { UserPage } from './UserPage'
import { AppContext } from '../../contexts/AppContext'
import * as api from '../../helpers/api';
import { MemoryRouter, Route } from 'react-router-dom'
import { act } from 'react-dom/test-utils';

jest.mock('../../helpers/api');
const mockHistoryGoBack = jest.fn();
const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        goBack: mockHistoryGoBack,
        push: mockHistoryPush
    })
}));

const renderWithContextAndRouterForExistingUser = (component) => {
    api.doGet.mockResolvedValue({
        id: 2,
        name: 'user',
        isAdmin: false
    });
    const context = { auth: { info: {} } };
    return {
        ...render(
            <AppContext.Provider value={context}>
                <MemoryRouter initialEntries={[`/users/2/edit`]}>
                    <Route path="/users/:userId/edit">
                        {component}
                    </Route>
                </MemoryRouter>
            </AppContext.Provider>)
    }
}

const renderWithContextAndRouterForNewUser = (component) => {
    const context = { auth: { info: {} } };
    return {
        ...render(
            <AppContext.Provider value={context}>
                <MemoryRouter initialEntries={[`/users/new`]}>
                    <Route path="/users/new">
                        {component}
                    </Route>
                </MemoryRouter>
            </AppContext.Provider>)
    }
}

afterEach(cleanup)

it('should match the snapshot for an existing user', async () => {
    let fragment;
    await act(async () => {
        const { asFragment } = renderWithContextAndRouterForExistingUser(<UserPage />);
        fragment = asFragment;
    });
    expect(fragment(<UserPage />)).toMatchSnapshot();
});

it('should match the snapshot for a new user', async () => {
    const { asFragment } = renderWithContextAndRouterForNewUser(<UserPage />);
    expect(asFragment(<UserPage />)).toMatchSnapshot();
});

it('should allow delete an existing user', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouterForExistingUser(<UserPage />);
    });

    await wait(() => {
        fireEvent.click(container.getByTestId('delete'));
    })

    expect(mockHistoryPush.mock.calls.length).toBe(1);
    expect(mockHistoryPush.mock.calls[0][0]).toBe('/users/2/delete');
    mockHistoryPush.mockClear();
});

it('should allow cancel', async () => {
    const { getByTestId } = renderWithContextAndRouterForNewUser(<UserPage />);

    await wait(() => {
        fireEvent.click(getByTestId('cancel'));
    })

    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});

it('should require user name', async () => {
    const { getByTestId } = renderWithContextAndRouterForNewUser(<UserPage />);

    await wait(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(getByTestId('userNameErrors')).toHaveTextContent('Required');
})

it('should update an existing user', async () => {
    let container
    await act(async () => {
        container = renderWithContextAndRouterForExistingUser(<UserPage />);
    });

    await changeInputValue(container.getByTestId, 'name', 'updated user');
    await changeInputValue(container.getByTestId, 'isAdmin', true);
    await changeInputValue(container.getByTestId, 'newPassword', 'pass');
    await changeInputValue(container.getByTestId, 'confirmNewPassword', 'pass');

    api.doPut.mockResolvedValue({});

    await wait(() => {
        fireEvent.click(container.getByTestId('submit'));
    })

    expect(api.doPut.mock.calls.length).toBe(1);
    expect(api.doPut.mock.calls[0][0]).toBe('users/2');
    expect(api.doPut.mock.calls[0][1]).toStrictEqual({ name: 'updated user', isAdmin: true, newPassword: 'pass', confirmNewPassword: 'pass' });

    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});

it('should create a new user', async () => {
    const { getByTestId } = renderWithContextAndRouterForNewUser(<UserPage />);

    await changeInputValue(getByTestId, 'name', 'new user');
    await changeInputValue(getByTestId, 'newPassword', 'pass');
    await changeInputValue(getByTestId, 'confirmNewPassword', 'pass');

    api.doPost.mockResolvedValue({});

    await wait(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(api.doPost.mock.calls.length).toBe(1);
    expect(api.doPost.mock.calls[0][0]).toBe('users');
    expect(api.doPost.mock.calls[0][1]).toStrictEqual({ name: 'new user', isAdmin: false, newPassword: 'pass', confirmNewPassword: 'pass' });

    expect(mockHistoryGoBack.mock.calls.length).toBe(1);
    mockHistoryGoBack.mockClear();
});

const changeInputValue = async (getByTestId, name, value) => {
    await wait(() => {
        fireEvent.change(getByTestId(name), {
            target: {
                value: value
            }
        })
    })
}
