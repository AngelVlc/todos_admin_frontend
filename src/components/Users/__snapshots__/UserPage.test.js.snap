// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot for a new user 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      New user
    </h3>
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/users"
          >
            Users
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/users/undefined"
          >
            new
          </a>
        </li>
      </ul>
    </nav>
    <form
      action="#"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="name"
        >
          Name
        </label>
        <div
          class="control"
        >
          <input
            data-testid="name"
            name="name"
            type="text"
            value=""
          />
        </div>
        <p
          class="help is-danger"
          data-testid="userNameErrors"
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="password"
        >
          Password
        </label>
        <div
          class="control"
        >
          <input
            data-testid="password"
            name="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="confirmPassword"
        >
          Confirm Password
        </label>
        <div
          class="control"
        >
          <input
            data-testid="confirmPassword"
            name="confirmPassword"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="isAdmin"
        >
          Is Admin
        </label>
        <div
          class="select"
        >
          <select
            data-testid="isAdmin"
            name="isAdmin"
          >
            <option
              value="yes"
            >
              Yes
            </option>
            <option
              value="no"
            >
              No
            </option>
          </select>
        </div>
      </div>
      <div
        class="field is-grouped"
      >
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="submit"
            type="submit"
          >
            CREATE
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="cancel"
            type="button"
          >
            CANCEL
          </button>
        </p>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should match the snapshot for an existing admin user 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      Edit admin user 'user'
    </h3>
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/users"
          >
            Users
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/users/2"
          >
            user
          </a>
        </li>
      </ul>
    </nav>
    <form
      action="#"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="name"
        >
          Name
        </label>
        <div
          class="control"
        >
          <input
            data-testid="name"
            name="name"
            type="text"
            value="user"
          />
        </div>
        <p
          class="help is-danger"
          data-testid="userNameErrors"
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="password"
        >
          Password
        </label>
        <div
          class="control"
        >
          <input
            data-testid="password"
            name="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="confirmPassword"
        >
          Confirm Password
        </label>
        <div
          class="control"
        >
          <input
            data-testid="confirmPassword"
            name="confirmPassword"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="isAdmin"
        >
          Is Admin
        </label>
        <div
          class="select"
        >
          <select
            data-testid="isAdmin"
            name="isAdmin"
          >
            <option
              value="yes"
            >
              Yes
            </option>
            <option
              value="no"
            >
              No
            </option>
          </select>
        </div>
      </div>
      <div
        class="field is-grouped"
      >
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="submit"
            type="submit"
          >
            SAVE
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="cancel"
            type="button"
          >
            CANCEL
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button is-danger"
            data-testid="delete"
            type="button"
          >
            DELETE
          </button>
        </p>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should match the snapshot for an existing non admin user 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      Edit user 'user'
    </h3>
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/users"
          >
            Users
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/users/2"
          >
            user
          </a>
        </li>
      </ul>
    </nav>
    <form
      action="#"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="name"
        >
          Name
        </label>
        <div
          class="control"
        >
          <input
            data-testid="name"
            name="name"
            type="text"
            value="user"
          />
        </div>
        <p
          class="help is-danger"
          data-testid="userNameErrors"
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="password"
        >
          Password
        </label>
        <div
          class="control"
        >
          <input
            data-testid="password"
            name="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="confirmPassword"
        >
          Confirm Password
        </label>
        <div
          class="control"
        >
          <input
            data-testid="confirmPassword"
            name="confirmPassword"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="isAdmin"
        >
          Is Admin
        </label>
        <div
          class="select"
        >
          <select
            data-testid="isAdmin"
            name="isAdmin"
          >
            <option
              value="yes"
            >
              Yes
            </option>
            <option
              value="no"
            >
              No
            </option>
          </select>
        </div>
      </div>
      <div
        class="field is-grouped"
      >
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="submit"
            type="submit"
          >
            SAVE
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="cancel"
            type="button"
          >
            CANCEL
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button is-danger"
            data-testid="delete"
            type="button"
          >
            DELETE
          </button>
        </p>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
